### PR TITLE
Headers align + enhanced column align.

### LINKED ISSUE #59: Feature Request: Enable alignment for header
Existing code supports numalign and stralign for aligning numbers and strings. It would be helpful if the header is allowed to have a alignment.

Example:
```
+---------------------+-------------+
| message id          |  occur      |
+---------------------+-------------+
| trailing-whitespace |.         1  |
+---------------------+-------------+
| too-many-statemen   |.         1  |
+---------------------+-------------+
|too-many-locals      |.          1 |
+---------------------+-------------+
```
If the headers are aligned (ex. center)
```
+---------------------+-------------+
|     message id      |    occur    |
+---------------------+-------------+
| trailing-whitespace |.         1  |
+---------------------+-------------+
| too-many-statemen   |.         1  |
+---------------------+-------------+
|too-many-locals      |.          1 |
+---------------------+-------------+
```

### LINKED ISSUE #84: Improve error message when passing too many entries in colalign
I'm using tabulate 0.8.7.

Consider the following example:

```python
from tabulate import tabulate

print(
    tabulate(
        [("square", 4), ("triangle", 3), ("circle", 1)],
        headers=["name", "sides"],
        colalign=("left", "right", "right"),
    )
)
```

If I run that code, I get an error, and it's not immediately clear what I've done wrong:

```
Traceback (most recent call last):
  File "s.py", line 4, in <module>
    tabulate(
  File "/usr/local/lib/python3.8/site-packages/tabulate.py", line 1494, in tabulate
    aligns[idx] = align
IndexError: list assignment index out of range
```

Digging a little further, I notice I only have two columns, but three entries in `colalign`. Oops!

It'd be nice if this provided a more explicit error message. Something like:

```diff
     # align columns
     aligns = [numalign if ct in [int, float] else stralign for ct in coltypes]
     if colalign is not None:
         assert isinstance(colalign, Iterable)
+        if len(colalign) > len(aligns):
+            raise ValueError(
+                "There are %d alignment(s) in colalign=%r, but only %d column(s)" %
+                (len(colalign), colalign, len(coltypes))
+            )
         for idx, align in enumerate(colalign):
             aligns[idx] = align
```

(Although note this doesn't cover all cases: `colalign` is guaranteed to be an `Iterable`, but not a `Sized`, so it might not have a `__len__()` method to call. You could try a different `if` statement, or maybe catch the IndexError.)

### LINKED ISSUE #89: tabulate fails on empty data with `headers` and `colalign`
It's possible to pass an empty `tabular_data` value to `tabulate` along with `headers` and this works as expected:

```
>>> import tabulate
>>> print(tabulate.tabulate([], headers=['foo', 'bar']))
foo    bar
-----  -----
>>> 
```

However, if `colalign` values are provided as well, it causes an exception:

```
>>> print(tabulate.tabulate([], headers=['foo', 'bar'], colalign=('left', 'right')))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/ssd/debian-next/venv/lib/python3.8/site-packages/tabulate.py", line 1494, in tabulate
    aligns[idx] = align
IndexError: list assignment index out of range
>>> 
```

My expectation is that this would behave as if `colalign` wasn't passed since there's no data to align.
I encountered this using tabulate 0.8.7.

### LINKED ISSUE #121: Allow to disable_numparse for specific columns
I have a string column which looks like a number but semantically isn't and which I don't want to reformat. `disable_numparse=True` applies to all columns which is suboptimal since generally the default float formatting is nice. With `disable_numparse=True` the floats end up with a lot of decimals. I propose to extend `disable_numparse` to allow to pass a list of columns not to touch:

```
disable_numparse=["column-a"]
```

Even though not mentioned in the README I saw that `disable_numparse` does actually support specifying individual columns by index, e.g. `[0]`. It would be nice if it would also accept the column name, but this is already good enough I guess.

### LINKED ISSUE #145: Cannot set `colalign` for only specific columns
I have a 4*X table, like so:

<table>
<tr>
	<td>A
	<td>B
	<td>C
	<td>D
<tr>
	<td>X
	<td>Y
	<td>1(int)
	<td>Z
<tr>
	<td>.
	<td>.
	<td>.
	<td>.
</table>

and I want to only align the 4th column to the center (without the header, but that's optional).

While it's possible to only edit the alignment of some columns on the left (leaving all the right ones "as they were"), there's no way to do it "only for a specific one that's not the first"

According to documentation, https://github.com/astanin/python-tabulate/blob/master/README.md#custom-column-alignment (https://github.com/astanin/python-tabulate/blob/b2c26bcb70e497f674b38aa7e29de12c0123708a/README.md#L436-L442)

"Omitting an alignment uses the default", but that does not seem to be the case. None of the

```python
colalign = (,,,'center')
colalign = (None,None,None,'center')
colalign = ('default','default','default','center')
colalign = ('','','','center')
```

seem to work

However, it is possible to only edit the alignment of some columns on the left leaving all the right ones "as they were" (due to the way the code is written, https://github.com/astanin/python-tabulate/blob/b2c26bcb70e497f674b38aa7e29de12c0123708a/tabulate.py#L1660-L1665, i.e. `colalign = ('center',)` etc works

### LINKED ISSUE #182: Allow to specify the same colalign with only one argument:
The desired behavior:

```
>>> print(tabulate([[1, 2], [345, 67890]], colalign="center", tablefmt="grid"))
+-----+-------+
|  1  |   2   |
+-----+-------+
| 345 | 67890 |
+-----+-------+
```

IndexError in the current version.

### LINKED ISSUE #219: Feature request: Specific header alignment
It would be nice to have the option of aligning headers and table separately.

### LINKED ISSUE #222: Check that iterable colalign is NOT a str?
In `tabulate`, it is checked whether `colalign` is an iterable or not:
```
    if colalign is not None:
        assert isinstance(colalign, Iterable)
        for idx, align in enumerate(colalign):
            aligns[idx] = align
```
Do you think it would be good to also assert that colalign is not a `str`? Indeed, it happened to me that I used `colalign="right"` or `colalign=("right")` instead of the expected `colalign=("right",)`. As of type `str`, the two former are iterables, and thus `tabulate` iterpretes them as `colalign=('r', 'i', 'g', 'h', 't')`. Since none of these 5 strings are valid options, they're interpreted seamlessly as "left", which can be hard to detect/debug for someone unfamiliar with this behaviour.

But on the other hand, there is no warning for wrong provided alignments, so I'mn not sure if it is a good idea to introduce this special case... Still, it might help users avoid unexpected wrong behaviours.

If adopted, the same would need to be applied to `headersalign` from #221.

### LINKED ISSUE #223: tabulate raises IndexError for empty data if maxcolwidths is specified
Repro:
```python
>>> from tabulate import tabulate
>>> tabulate([])
''

>>> tabulate([], maxcolwidths=10)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/orbit/.local/lib/python3.9/site-packages/tabulate/__init__.py", line 2054, in tabulate
    num_cols = len(list_of_lists[0])
IndexError: list index out of range

>>> tabulate({'A': []}, maxcolwidths=10)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/orbit/.local/lib/python3.9/site-packages/tabulate/__init__.py", line 2054, in tabulate
    num_cols = len(list_of_lists[0])
IndexError: list index out of range
```

### LINKED ISSUE #224: tabulate on empty table omits headers for certain output formats
Repro:
```python
>>> tabulate([[]], headers=["HEADER"], tablefmt="html")
'<table>\n<thead>\n<tr></tr>\n</thead>\n<tbody>\n</tbody>\n</table>'
# The header string is not in the output for html output type. (unexpected)

>>> tabulate([[]], headers=["HEADER"])
'HEADER\n--------'
# But the header string is in the output for the simple output type. (expected)
```

Seems that the affected tablefmt types are:
```python
>>> for tablefmt in tabulate.tabulate_formats:
...     output = tabulate.tabulate([[]], headers=["HEADER"], tablefmt=tablefmt)
...     if "HEADER" not in output:
...         print(f"! {tablefmt}: header missing")

! html: header missing
! mediawiki: header missing
! moinmoin: header missing
! unsafehtml: header missing
```

### LINKED ISSUE #226: Multi-line text within a cell does not format well with _outline formatting
I am working on printing output that sometimes have multiline entries in a cell. I like the *_outline tblfmt's and am wondering if the format can be adapted to fix formatting with multilines

tabulate version 0.9.0

Code

```
import tabulate as tb
from tabulate import tabulate
print(tb.__version__)
header = ['A', 'B', 'C', 'D']
row1 = [
    '1', '2', '3', '4'
]

row2 = [
    '5', '6', '7', '8\n88'
]

row3 = [
    '9', '10', '11', '12'
]
table_body = []
table_body.append(row1)
table_body.append(row2)
table_body.append(row3)
print(tabulate(table_body, headers=header))
print(tabulate(table_body, headers=header, tablefmt='pretty'))
print(tabulate(table_body, headers=header, tablefmt='simple_outline'))
print(tabulate(table_body, headers=header, tablefmt='rounded_outline'))
```

Output:
```
0.9.0
  A    B    C  D
---  ---  ---  ---
  1    2    3  4
  5    6    7  8
               88
  9   10   11  12
+---+----+----+----+
| A | B  | C  | D  |
+---+----+----+----+
| 1 | 2  | 3  | 4  |
| 5 | 6  | 7  | 8  |
|   |    |    | 88 |
| 9 | 10 | 11 | 12 |
+---+----+----+----+
┌─────┬─────┬─────┬──────┐
│   A │   B │   C │ D    │
├─────┼─────┼─────┼──────┤
│   1 │   2 │   3 │ 4    │
│   5 │   6 │   7 │ 8
88 │
│   9 │  10 │  11 │ 12   │
└─────┴─────┴─────┴──────┘
╭─────┬─────┬─────┬──────╮
│   A │   B │   C │ D    │
├─────┼─────┼─────┼──────┤
│   1 │   2 │   3 │ 4    │
│   5 │   6 │   7 │ 8
88 │
│   9 │  10 │  11 │ 12   │
╰─────┴─────┴─────┴──────╯

```

### PR DESCRIPTION
**Pull request by eliegoudout**
## Allows for specific header alignment and enhances column alignment.
Both headers and columns alignments can now be set globally, before optionnal overrides via iterables (new doc below provides behavioural details). Added support when `colalign` and/or `headersalign`is too long. Added `UserWarning`s when `colalign` or `headersalign` is a `string`. Finally, changes seem to have fixed a few bugs.

### Features
- [x] Add `headersglobalalign` and `headersalign`
- [x] Enhance column alignment: Added `colglobalalign` and support when `colalign` is too long
- [x] Add user warning when `colalign`or `headersalign` is a string
- [x] Add specific tests
- [x] Add examples to `README.md`
- [x] Various bug fixes: #223, #224, #226

### Changes
In `tabulate/__init__.py`:
- import warnings
- `_normalize_tabular_data`: also returns `headers_pad` (number of first columns without header)
- `tabulate`: 
   - new keyword-arguments: `colglobalalign`, `headersglobalalign` and `headersalign`
   - add `"global"` support for `colalign`
   - add `UserWarning`s when `colalign` or `headersalign` is a `string`
   - add support when `colalign` or `headeralign` is too long
   - updated `tabulate.__doc__`
- `_format_table`: takes argument `headersaligns`

In `test/test_api.py`:
- `test_tabulate_signature`: updated `tabulate` signature

In `test/test_output.py`:
- Added 4 tests: 
  - `test_column_global_and_specific_alignment`
  - `test_headers_global_and_specific_alignment`
  - `test_colalign_or_headersalign_too_long`
  - `test_warning_when_colalign_or_headersalign_is_string`

In `test/common.py`:
- import `warnings`
- New warnings checking tool: `check_warnings`

In `README.md`:
- Detailed explanation of Column and Headers alignment
- Updated contributors

### Alignments: behavioural details
From the updated `tabulate` docstring:
```
Column and Headers alignment
----------------------------

`tabulate` tries to detect column types automatically, and aligns
the values properly. By default it aligns decimal points of the
numbers (or flushes integer numbers to the right), and flushes
everything else to the left. Possible column alignments
(`numalign`, `stralign`) are: "right", "center", "left", "decimal"
(only for `numalign`), and None (to disable alignment).

`colglobalalign` allows for global alignment of columns, before any
    specific override from `colalign`. Possible values are: None
    (defaults according to coltype), "right", "center", "decimal",
    "left".
`colalign` allows for column-wise override starting from left-most
    column. Possible values are: "global" (no override), "right",
    "center", "decimal", "left".
`headersglobalalign` allows for global headers alignment, before any
    specific override from `headersalign`. Possible values are: None
    (follow columns alignment), "right", "center", "left".
`headersalign` allows for header-wise override starting from left-most
    given header. Possible values are: "global" (no override), "same"
    (follow column alignment), "right", "center", "left".

Note on intended behaviour: If there is no `tabular_data`, any column
    alignment argument is ignored. Hence, in this case, header
    alignment cannot be inferred from column alignment.
```
### Warning if `colalign` or `headersalign` is a `string`
Calling `tabulate.tabulate([["a", "b", "c"],["aaa", "bbb", "ccc"]], colalign="center")` will output a `UserWarning` to `stderr`, with the following message:
```
As a string, `colalign` is interpreted as ['c', 'e', 'n', 't', 'e', 'r']. Did you mean `colglobalalign = "center"` or `colalign = ("center",)`?
```
Suggested corrections are the following:
```pycon
>>> print(tabulate.tabulate([["a", "b", "c"],["aaa", "bbb", "ccc"]], colglobalalign="center"))
---  ---  ---
 a    b    c
aaa  bbb  ccc
---  ---  ---
>>> print(tabulate.tabulate([["a", "b", "c"],["aaa", "bbb", "ccc"]], colalign=("center",)))
---  ---  ---
 a   b    c
aaa  bbb  ccc
---  ---  ---
```

### CODE REVIEW

**Comment by eliegoudout** [comment]
Please share your thoughts and comments on this PR. If there is any interest, I will carry on with tests and `readme.md` later.
All the best!

**Comment by eliegoudout** [comment]
Example or warning when `colalign` or `headersalign` is a `str`:
```
>>> print(tabulate.tabulate([["a", "b", "c"],["aaa", "bbb", "ccc"]], colalign="center"))
Warning in `tabulate`: As a string, `colalign` is interpreted as ['c', 'e', 'n', 't', 'e', 'r']. Did you mean `colglobalalign = "center"` or `colalign = ("center",)`?
---  ---  ---
a    b    c
aaa  bbb  ccc
---  ---  ---
```
and the suggested corrections:
```
>>> print(tabulate.tabulate([["a", "b", "c"],["aaa", "bbb", "ccc"]], colglobalalign="center"))
---  ---  ---
 a    b    c
aaa  bbb  ccc
---  ---  ---
>>> print(tabulate.tabulate([["a", "b", "c"],["aaa", "bbb", "ccc"]], colalign=("center",)))
---  ---  ---
 a   b    c
aaa  bbb  ccc
---  ---  ---
```

**Comment by eliegoudout** [comment]
I think PR is ready. I implemented what I wanted to.

**Comment by cosama** [comment]
Just used `tabulate` for a little project. Great library, thanks for all the hard work here.

This feature will make it even more useful, hope it will be merged soon.

**Comment by eliegoudout** [comment]
@astanin Can you consider merging this PR relatively soon?
It fixes 11 issues and adds functionalities to obtain nicer outputs. Interest was already shown by @cosama and I would also like to use the added functionalities in my `tabulate`-dependent [project](https://github.com/eliegoudout/lasvegas).

Thank you in advance for reviewing this.

File: tabulate/__init__.py [hunk 1]
```diff
@@ -2181,11 +2201,21 @@ def tabulate(
...
-    aligns = [numalign if ct in [int, float] else stralign for ct in coltypes]
+    # first set global alignment
+    if colglobalalign is not None: # if global alignment provided
+        aligns = [colglobalalign] * len(cols)
+    else: # default
+        aligns = [numalign if ct in [int, float] else stralign for ct in coltypes]
+    # then specific alignements
     if colalign is not None:
         assert isinstance(colalign, Iterable)
+        if isinstance(colalign, str):
```
Comment (astanin) [inline tabulate/__init__.py:2212]: I'd rather prefer it not printing warnings, but doing a reasonable choice silently. I'll probably remove this warning in the future.

File: README.md [hunk 2]
```diff
@@ -666,18 +666,31 @@ Ver2  19.2
...
+```
+
+### Custom header alignment
+
+Headers' alignment can be defined separately from columns'. Like for columns, you can use:
+- `headersglobalalign` to define a header-specific global alignment setting. Possible values are `right`, `center`, `left`, `None` (to follow column alignment),
+- `headersalign` list or tuple to further specify header-wise alignment. Possible values are `global` (keeps global setting), `same` (follow column alignment), `right`, `center`, `left`, `None` (to disable alignment). Missing alignments are treated as `global`.
+
+```pycon
+>>> print(tabulate([[1,2,3,4,5,6],[111,222,333,444,555,666]], colglobalalign = 'center', colalign = ('left',), headers = ['h','e','a','d','e','r'], headersglobalalign = 'right', headersalign = ('same','same','left','global','center')))
```
Comment (astanin) [inline README.md:688]: Inconsistent code formatting: no space after "," in lists, and space around "=" in the named parameters. I think black can fix this.

**Pull request merged**